ChronoStore/src/storyreader.cpp

```diff
@@ -0,0 +1,320 @@
+        free(attributeDataBuffer);
+        if(story >= 0)    H5Dclose(story);
+        if(chronicle >= 0)    H5Fclose(chronicle);
+        dr.status = -1;
+        // std::cout<<"Min: "<<dmm.MinMax[0]<<" Max: "<<dmm.MinMax[1]<<" "<<dmm.status<<std::endl;
+        return dr; 
+    }
+
+    /*
+     * Find the chunks to read
+     */
+    std::pair<int64_t, int64_t> chunksToRead = {INT_MIN , INT_MIN};
+    if(attributeDims == 1){
+        chunksToRead.first = 0;
+        chunksToRead.second = 0;
+    }
+    else{
+        chunksToRead.first = searchFirstChunk(attributeDataBuffer, range.first, attributeDims - 1, 0);
+        if(chunksToRead.first != INT_MIN){
+            chunksToRead.second = searchLastChunk(attributeDataBuffer, range.second, attributeDims - 1, chunksToRead.first);
+        }
+    }
+
+    // std::cout<<"Chunks to read: "<<chunksToRead.first<<" "<<chunksToRead.second<<std::endl;
+    /*
+     * Check for the range of requested query of chunks whether the chunkToRead is valid
+     */
+    if(chunksToRead.second == INT_MIN || chunksToRead.first == INT_MIN){
+        if(attributeDataBuffer != NULL) free(attributeDataBuffer);
+        if(story >= 0)    H5Dclose(story);
+        if(chronicle >= 0)    H5Fclose(chronicle);
+        return dr;
+    }
+
+    /*
+     *  Declare eventType
+     */
+    herr_t timeStampError, dataError;
+    eventType = H5Tcreate(H5T_COMPOUND, sizeof(Event));
+    timeStampError = H5Tinsert(eventType, "timeStamp", HOFFSET(Event, timeStamp), H5T_NATIVE_ULLONG);
+    dataError = H5Tinsert(eventType, "data", HOFFSET(Event, data), H5T_NATIVE_CHAR);
+
+    if(eventType < 0 || dataError < 0 || timeStampError < 0){
+        if(eventType >=0 )  H5Tclose(eventType);
+        if(attributeDataBuffer != NULL) free(attributeDataBuffer);
+        if(story >= 0)  H5Dclose(story);
+        if(chronicle >= 0)  H5Fclose(chronicle);
+        return dr;
+    }
+
+    hsize_t datasetHyperslabStart[1] = {0};
+    hsize_t storyChunkDims[1];
+    herr_t memSpaceErr, datasetSpaceError;
+    hid_t hyperslabMemorySpace;
+    storyDatasetSpace  = H5Dget_space(story);
+    hsize_t chunkOffset[1] = {0};
+    try{
+
+        /* 
+        * Create a memory storyDatasetSpace to hold the chunk data 
+        */
+        for(int chunkIndex = chunksToRead.first; chunkIndex <= chunksToRead.second ; chunkIndex++){
+            storyChunkDims[0] = {attributeDataBuffer[chunkIndex].totalChunkEvents};
+            chunkOffset[0] = {attributeDataBuffer[chunkIndex].datasetIndex};
+
+            hyperslabMemorySpace = H5Screate_simple(1, storyChunkDims, NULL);
+            memSpaceErr = H5Sselect_hyperslab(hyperslabMemorySpace, H5S_SELECT_SET, datasetHyperslabStart, NULL, storyChunkDims, NULL);
+            datasetSpaceError = H5Sselect_hyperslab(storyDatasetSpace, H5S_SELECT_SET, chunkOffset, NULL, storyChunkDims, NULL);
+
+            eventBuffer = (Event*) malloc(attributeDataBuffer[chunkIndex].totalChunkEvents * sizeof(Event));
```

Comment:
Use STL vectors and new for eventBuffer.

Reply:
Done